Update moshi in library and sample app and use new way of using sealed types

daniellevass opened the pull request:
## What
On of our dependencies, Moshi, changed how sealed types and our polymorphic way of reading data for our server sync events. 

From https://www.zacsweers.dev/a-closer-look-at-moshi-1-9/
> In projects using code gen, there are cases where Kotlin classes could have (appeared to) Just Work™️ before if you forgot to annotate them with @JsonClass. These will fail at runtime now.

We needed to follow the instructions Zac's written in https://github.com/ZacSweers/moshi-sealed to make our library work work in the new system. So I've added those dependencies and followed the new ways.

## Why
Users could experience a crash if they updated their own moshi dependency to 1.9.x - we shouldn't crash if they update. This should fix #100

marekoid commented on pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt:
```diff
@@ -1,35 +1,44 @@
 
 // If you add new fields to any of these data classes, make sure they have
 // default values otherwise they can be null when reading the old stored
 // JSON from disk.
```
That comment have been no longer valid already before the changes in this PR but now it's a good opportunity to delete it.

luismfonseca replied:
@daniellevass can you move some of your PR comment you did [here](https://github.com/pusher/push-notifications-android/pull/101#issuecomment-567611532) to this comment.

daniellevass replied:
Have updated the comment to point people to the tests for the single source of truth! I updated the tests to more explicitly detail what results are expected and added comments there if data is expected not to be returned e.g. if a field is missing. Does that make sense?

marekoid commented on pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt:
```diff
@@ -1,35 +1,44 @@
-                .withSubtype(StopJob::class.java, "StopJob")
-    }
-}
+@JsonClass(generateAdapter = true, generator = "sealed:ServerSyncJob")
```
It does not seem like we're using code gen here (no kapt dependency). Those annotations here look redundant.

marekoid replied:
Or rather I hope that all that boiler-plate is redundant ;) Definitely the annotation values about generation seem redundant.

daniellevass replied:
Yeah, I think you were right! I removed almost all of them and everything works correctly. I think the one on the sealed type is still necessary as it tells everything else it's a sealed type. Does that make sense?

marekoid replied:
Awesome, thanks Danielle! I've checked that making the annotation that needs to stay not conveying information that code gen is involved (whilst it isn't), works (at least the tests involved in this diff are green on my machine).
```suggestion
@JsonClass(generateAdapter = false, generator = "sealed:ServerSyncJob")
```
Well it still sounds a bit like code gen is involved ("generator") but not that explicitly and hopefully `generateAdapter = false` clearly conveys that it is not. It feels though like an imperfection of that super early preview of the library extension so that's probably the best what we can achieve now.

daniellevass replied:
That sounds fair! I've my code to reflect that.

marekoid commented on pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt:
```diff
@@ -1,35 +1,44 @@
-    }
-}
+@JsonClass(generateAdapter = true, generator = "sealed:ServerSyncJob")
+sealed class ServerSyncJob: Serializable { }
```
`Serializable` marker looks redundant and if it is indeed redundant it is also confusing and error-prone.

daniellevass replied:
hmm. the code doesn't compile if I remove this. The ServerSyncHandler has a JobQueue ( which expects the types to be Serializable (https://github.com/pusher/push-notifications-android/blob/15f253d5ade2d783cd8aae52763de3c0bbcfcd11/pushnotifications/src/main/java/com/pusher/pushnotifications/internal/PersistentJobQueue.kt). I think this would require more refactoring to remove.

marekoid replied:
Oh I didn't expect serialisation is being done using two mechanisms. Thanks taking the feedback onboard and for checking! Yeah refactoring of that is not valuable enough to give it a shot.

marekoid commented on sample_kotlin/build.gradle:
```diff
@@ -43,6 +43,8 @@ dependencies {
     implementation 'com.google.firebase:firebase-core:16.0.9'
     implementation 'com.google.firebase:firebase-messaging:18.0.0'
+
+    implementation 'com.squareup.moshi:moshi-kotlin:1.9.2'
```
Why do we need to add moshi dependency to the sample_kotlin module?

luismfonseca replied:
Yes, I'd expect things to work without this, so I think it's best if we don't include it here.

daniellevass replied:
I've removed this, it was useful to test that the demo app still worked correctly with this dependency (previously the app would crash).

marekoid commented on pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt:
```diff
@@ -1,35 +1,44 @@
 // JSON from disk.
+@TypeLabel("StartJob")
+@JsonClass(generateAdapter = true)
 data class StartJob(val fcmToken: String, val knownPreviousClientIds: List<String>): ServerSyncJob()
```
How about the migration? Won't we have a similar issue like the last time since we're changing how this sealed class hierarchy is unmarshalled? Won't there be any jobs persisted by a version prior to these changes of updating Moshi?

daniellevass replied:
Yeah. We'll have difficulties migrating if any of the underlying classes change. As we've not changed any of the structure of those classes everything seems to be happy! I've updated the comment at the top of `ServerSyncJob` to explain what will happen if those classes change, and explained that the `ServerSyncJobTests` are our single source of truth for what's expected to happen :-)

marekoid replied:
Cool thanks for checking that!

marekoid commented on pushnotifications/build.gradle:
```diff
@@ -32,8 +32,10 @@ dependencies {
+    implementation 'com.squareup.moshi:moshi-kotlin:1.9.2'
+    implementation 'com.squareup.moshi:moshi-adapters:1.9.2'
+    implementation "dev.zacsweers.moshisealed:moshi-sealed-reflect:0.1.0"
+    implementation "dev.zacsweers.moshisealed:moshi-sealed-annotations:0.1.0"
```
Not sure if the solution shouldn't be not to marshal a sealed class if it's not supported by Moshi... That extension lib created by one of the contributors is [WIP even when it comes to its API](https://www.reddit.com/r/androiddev/comments/dpjoy4/moshisealed_reflective_and_code_gen/f5vwsgn/). Also it seems like even with that lib, we need to remember about a lot of boiler-plate to put at each subclass of the sealed class.

Alternative solution could be about separately persisting given types of `ServerSyncJob`s so that we always know which type to expect when unmarshalling.

daniellevass replied:
I think this is a question that @luismfonseca could help answer - are we okay using a WIP library to mitigate using sealed classes with moshi, or we could spend some time refactoring this piece of code to explicitly save the classes and not use a sealed type?

marekoid replied:
From my side that comment is more to serve as lesson learned for new code, sorry for not expressing it clearly. I don't think it would be a good ROI in doing a broader and tricky refactoring here.

luismfonseca replied:
> are we okay using a WIP library to mitigate using sealed classes with moshi, or we could spend some time refactoring this piece of code to explicitly save the classes and not use a sealed type?

I think if we do sufficient testing on our end to make sure the library works well for us, and if we are happy with how it's performing, then we can use it. The fact that's not an official library means that we need to put in more effort to make sure it works for us.

Also, I think that ideally we have the discussions about the way to solve something should happen before a PR -- not always possible, but how we go about solving something should be discussed with the potential reviewers, minimizing the late "this is better if implemented in this completely different way".

daniellevass replied:
Yeah, that sounds fair! I'll test on the sample app again to be 100% certain this is safe for us to use.

Totally agree with discussing approaches before PRs - in future I'll set up a huddle before finalising my approach to discuss any hesitations anyone may have before starting the work. Does that work?

daniellevass commented:
Marek left some feedback that we should double check that this works okay with people who already have a queue and will be upgrading, whether the core data itself would work. 

I thought we had tests that did this, and it turns out we do in `PersistentJobQueueTest` - we have three saved files that we access directly. However, they were failing 😱

Firstly, I've added a step in bitrise called `gradle test` which should run all our tests and actually notify us here if something has broken 🎉 

Secondly, it looks like the edge cases where the data models have changed is now different:

Previously if we had removed an attribute from an object that attribute became "null" or we needed to provide a default. Now that object will fail to parse :( 

Previously if we had added an attribute from an object, that attribute was ignored and the item was still returned. Now that object will also fail to parse :(

The current tests reflect the expected behaviour of using moshi 1.9 and have thusly been updated.

As the behaviour of this has changed I'm re-requesting reviews from you @luismfonseca -  I figure as we aren't likely to change the data models in the near term this shouldn't be an issue, but should be something you're aware of!

daniellevass commented:
I updated the tests to more clearly explain what's expected and what happens when it fails to parse a record. The source of truth should be the tests and not this PR :-)

Coincidentally, I'd noticed that the saved files didn't contain the records that I expected. So I re-ran the uncommented code to re-create the files.

marekoid commented:
Not directly related to this PR but it would be good to see test reports in Bitrise. That is what tests have been run to be sure all of them have been discovered.
